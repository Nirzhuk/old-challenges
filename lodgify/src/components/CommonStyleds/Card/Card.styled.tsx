
import styled from "styled-components";

export const CardStyled = styled.div`
    box-sizing: border-box;
    overflow: hidden;
    border-radius: 5px;
    background: #f7f7f7;
    box-shadow: 0 3px 5px rgba(0,0,0,5%);
    height: 280px;
    width: 220px;

`;

export const CardInnerContainer = styled.div`
    &:hover {
          border: inset 1px #333333;
    };
    height: 278px;
    border-radius: inherit;
`;
export interface FuelMeterProps {
  meter: number;
}

export const ShowcaseImageContainer = styled.div`
  position: relative;
  overflow: hidden;
  height: 140px;
  border-radius: 5px 5px 0 0;
`;
export const ShowcaseImageOverlay = styled.div`
  display: grid;
  grid-template-columns: 1fr;
  position: absolute;
  left: 0;
  z-index: 2;
`;
export const ShowcaseImage = styled.img`
  height: inherit;
  width: 100%;
`;
ShowcaseImage.defaultProps = {
  alt: "cardImage",
  onError: (ev: any) => { ev.target.onerror = null; ev.target.src = "/error-404.svg" }
}

export const ShowcaseContent = styled.div`
  display: grid;
  padding: 16px;
  grid-gap: 16px;
  height: 114px;
  grid-template-rows: max-content 1fr max-content;
`;

export const TitleSection = styled.div`
  display:grid;
  grid-template-rows: 1fr 1fr;
`;

export const CardTitle = styled.p`
  font-weight: 700;
  display: inline;
  margin: 0;
  font-size: 16px;
  font-style: italic;
`;
export const BottomSection = styled.div`

`;