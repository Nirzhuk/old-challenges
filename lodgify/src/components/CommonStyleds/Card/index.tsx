import { memo, useState, useEffect } from "react";
import Button from "../Button";
import Paragraph from "../Paragraph";
import Tag from "../Tag";
import {
    ShowcaseContent,
    CardStyled,
    ShowcaseImageContainer,
    ShowcaseImageOverlay,
    CardTitle,
    ShowcaseImage,
    CardInnerContainer,
    TitleSection,
    BottomSection
} from "./Card.styled";
import { CheckCircle } from 'react-feather'
import { BookableStatusType } from "../../../core/types/App";


interface CardProps {
    id: string;
    name: string;
    image: string;
    bookable: boolean;
    booked: number;
}


const Card = ({
    id,
    name,
    image,
    booked,
    bookable,
}: CardProps) => {

    const [bookableState, setBookableState] = useState({
        bookable: false,
        booked: 0
    })
    const handleClick = () => {
        setTimeout(function () {//simulate api
            setBookableState({
                bookable: true,
                booked: 1
            })
        }, 200);

    }

    const renderBottomSection = (booked: number, bookable: boolean) => {
        if (!bookable) {
            return <Button disabled>Not Bookable</Button>
        }
        if (booked > 0 && bookable) {
            return <Paragraph color="#11A960"> <CheckCircle size="16px" /> Booked for {booked} days</Paragraph>
        }
        else if (bookable) {
            return <Button onClick={handleClick}>Book</Button>
        }
    }
    const renderTag = (booked: number, bookable: boolean)  => {
        if (!bookable) {
            return <Tag state={BookableStatusType.UNAVAILABLE}>unavailable</Tag>
        }
        if (booked > 0 && bookable) {
            return <Tag state={BookableStatusType.BOOKED}>booked</Tag>
        }
        else if (bookable) {
            return <Tag state={BookableStatusType.AVAILABLE}>available</Tag>
        }


    }

    useEffect(() => {
        setBookableState({
            bookable,
            booked
        });
    }, [])//eslint-disable-line react-hooks/exhaustive-deps
    return (
        <CardStyled data-testid="card">
            <CardInnerContainer>
                <ShowcaseImageContainer>
                    <ShowcaseImageOverlay>
                        {renderTag(bookableState.booked, bookableState.bookable)}
                    </ShowcaseImageOverlay>
                    <ShowcaseImage loading="lazy" src={image} alt={name}/>
                </ShowcaseImageContainer>
                <ShowcaseContent>
                    <TitleSection>
                        <Paragraph as="span" color="#999999">Id: {id}</Paragraph>
                        <CardTitle>{name}</CardTitle>
                    </TitleSection>
                    <BottomSection>
                        {renderBottomSection(bookableState.booked, bookableState.bookable)}
                    </BottomSection>
                </ShowcaseContent>
            </CardInnerContainer>
        </CardStyled>
    );
};

export default memo(Card);
