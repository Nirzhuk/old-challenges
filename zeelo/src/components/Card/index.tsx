import { Book } from "modules/books/types";
import {
  CardContent,
  CardContentContainer,
  CardImage,
  CardImageContainer,
  CardStyled,
} from "./Card.styled";


interface CardProps {
    onClick(book:Book): void;
    book: Book;
}


const Card = ({
    onClick,
    book
}: CardProps) => {
    return (
        <CardStyled>
            <CardImageContainer>
                <CardImage onClick={() => onClick(book)} src={book.image} />
            </CardImageContainer>
            <CardContentContainer>
                <CardContent>
                    <strong>{book.title}</strong>
                </CardContent>
            </CardContentContainer>
        </CardStyled>
    );
};

export default Card;
