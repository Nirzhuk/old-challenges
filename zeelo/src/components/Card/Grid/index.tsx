import React from 'react'
import rangeMap from 'core/libs/utils/rangeMap';
import Card from '..';
import Skeleton from 'components/CommonStyleds/Skeleton';
import { Book } from 'modules/books/types';


interface GridCardProps {
    data: Book[]
    onClickCard(book:Book):void;
}

const CardGrid = ({ data, onClickCard }: GridCardProps) => {
    return <>{data && data.length !== 0 ? (
        data.map((book: any) => <Card
            key={book.id}
            onClick={onClickCard}
            book={book}
             />)
    ) : (rangeMap<React.ReactNode>(20, (i) => (
        <Skeleton
            height="350px"
            key={i}
        />
    )))}</>
};

export default React.memo(CardGrid)
