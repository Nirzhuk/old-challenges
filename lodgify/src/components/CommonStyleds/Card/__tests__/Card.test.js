import { render, screen } from '@testing-library/react'
import Card from '..';

test('Card Bookable', async () => {
    const houseData = {
        id: 1,
        name: 'House 1',
        image: 'https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png',
        booked: 0,
        bookable: true
    }
    const { container } = render(<Card {...houseData} />)
    expect(container.firstChild).toMatchSnapshot()
    expect(screen.queryByText("House 1")).toHaveTextContent("House 1")
    expect(screen.getByTestId('card').querySelector('img')).toHaveAttribute('src', houseData.image)
    expect(screen.getByTestId("button")).toHaveTextContent(`Book`)
})
test('Card Booked', async () => {
    const houseData = {
        id: 1,
        name: 'House 1',
        image: 'https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png',
        booked: 5,
        bookable: true
    }
    const { container } = render(<Card {...houseData} />)
    expect(container.firstChild).toMatchSnapshot()
    expect(screen.queryByText("House 1")).toHaveTextContent("House 1")
    expect(screen.getByTestId('card').querySelector('img')).toHaveAttribute('src', houseData.image)
    expect(screen.getByText('Booked for 5 days')).toHaveTextContent(`Booked for 5 days`)
})


test('Card Not Bookable', async () => {
    const houseData = {
        id:1,
        name:'House 1',
        image:'https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png',
        booked:0,
        bookable: false
    }
    const { container } = render(<Card {...houseData} />)
    expect(container.firstChild).toMatchSnapshot()
    expect(screen.queryByText("House 1")).toHaveTextContent("House 1")
    expect(screen.getByTestId('card').querySelector('img')).toHaveAttribute('src', houseData.image)
    expect(screen.getByTestId("button")).toHaveTextContent(`Not Bookable`)
})
