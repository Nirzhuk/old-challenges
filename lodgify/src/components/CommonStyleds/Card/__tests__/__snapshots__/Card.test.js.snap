// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Bookable 1`] = `
<div
  class="sc-hKFxyN dTzqPj"
  data-testid="card"
>
  <div
    class="sc-eCApnc igRcEp"
  >
    <div
      class="sc-jSFjdj betVMN"
    >
      <div
        class="sc-gKAaRy qBbhW"
      >
        <span
          class="sc-dlnjwi fyRIny"
          data-testid="tag"
        >
          available
        </span>
      </div>
      <img
        alt="House 1"
        class="sc-iCoGMd bTrehl"
        loading="lazy"
        src="https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png"
      />
    </div>
    <div
      class="sc-fujyAs bYSxkY"
    >
      <div
        class="sc-pNWdM hQjrQV"
      >
        <span
          class="sc-gtsrHT bbWVNj"
          color="#999999"
        >
          Id: 
          1
        </span>
        <p
          class="sc-jrsJWt Werog"
        >
          House 1
        </p>
      </div>
      <div
        class="sc-kEqXSa ixSTjh"
      >
        <button
          class="sc-bdnxRM fsLDke"
          data-testid="button"
        >
          Book
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Card Booked 1`] = `
<div
  class="sc-hKFxyN dTzqPj"
  data-testid="card"
>
  <div
    class="sc-eCApnc igRcEp"
  >
    <div
      class="sc-jSFjdj betVMN"
    >
      <div
        class="sc-gKAaRy qBbhW"
      >
        <span
          class="sc-dlnjwi iRZuWK"
          data-testid="tag"
        >
          booked
        </span>
      </div>
      <img
        alt="House 1"
        class="sc-iCoGMd bTrehl"
        loading="lazy"
        src="https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png"
      />
    </div>
    <div
      class="sc-fujyAs bYSxkY"
    >
      <div
        class="sc-pNWdM hQjrQV"
      >
        <span
          class="sc-gtsrHT bbWVNj"
          color="#999999"
        >
          Id: 
          1
        </span>
        <p
          class="sc-jrsJWt Werog"
        >
          House 1
        </p>
      </div>
      <div
        class="sc-kEqXSa ixSTjh"
      >
        <p
          class="sc-gtsrHT bzuYCR"
          color="#11A960"
        >
           
          <svg
            fill="none"
            height="16px"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
            />
            <polyline
              points="22 4 12 14.01 9 11.01"
            />
          </svg>
           Booked for 
          5
           days
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Card Not Bookable 1`] = `
<div
  class="sc-hKFxyN dTzqPj"
  data-testid="card"
>
  <div
    class="sc-eCApnc igRcEp"
  >
    <div
      class="sc-jSFjdj betVMN"
    >
      <div
        class="sc-gKAaRy qBbhW"
      >
        <span
          class="sc-dlnjwi jThYQV"
          data-testid="tag"
        >
          unavailable
        </span>
      </div>
      <img
        alt="House 1"
        class="sc-iCoGMd bTrehl"
        loading="lazy"
        src="https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png"
      />
    </div>
    <div
      class="sc-fujyAs bYSxkY"
    >
      <div
        class="sc-pNWdM hQjrQV"
      >
        <span
          class="sc-gtsrHT bbWVNj"
          color="#999999"
        >
          Id: 
          1
        </span>
        <p
          class="sc-jrsJWt Werog"
        >
          House 1
        </p>
      </div>
      <div
        class="sc-kEqXSa ixSTjh"
      >
        <button
          class="sc-bdnxRM fsLDke"
          data-testid="button"
          disabled=""
        >
          Not Bookable
        </button>
      </div>
    </div>
  </div>
</div>
`;
